server/src/main/java/org/elasticsearch/transport/OutboundHandler.java

```diff
@@ -200,39 +208,50 @@ void sendErrorResponse(
 
     private void sendMessage(
         TcpChannel channel,
+        @Nullable String requestAction,
         Writeable writeable,
-        OutboundMessage networkMessage,
+        long requestId,
+        boolean isHandshake,
+        Compression.Scheme compressionScheme,
+        TransportVersion version,
         ResponseStatsConsumer responseStatsConsumer,
         Releasable onAfter
     ) throws IOException {
-        final RecyclerBytesStreamOutput byteStreamOutput;
-        boolean bufferSuccess = false;
-        try {
-            byteStreamOutput = new RecyclerBytesStreamOutput(recycler);
-            bufferSuccess = true;
-        } finally {
-            if (bufferSuccess == false) {
-                Releasables.closeExpectNoException(onAfter);
-            }
-        }
+        compressionScheme = writeable instanceof BytesTransportRequest ? null : compressionScheme;
         final BytesReference message;
         boolean serializeSuccess = false;
+        final boolean isError = writeable instanceof RemoteTransportException;
+        final RecyclerBytesStreamOutput byteStreamOutput = new RecyclerBytesStreamOutput(recycler);
         try {
-            message = networkMessage.serialize(writeable, byteStreamOutput);
+            message = serialize(
+                requestAction,
+                requestId,
+                isHandshake,
+                version,
+                isError,
+                compressionScheme,
+                writeable,
+                threadPool.getThreadContext(),
+                byteStreamOutput
+            );
             serializeSuccess = true;
         } catch (Exception e) {
-            logger.warn(() -> "failed to serialize outbound message [" + networkMessage + "]", e);
+            logger.warn(() -> "failed to serialize outbound message [" + writeable + "]", e);
             throw e;
         } finally {
             if (serializeSuccess == false) {
                 Releasables.close(byteStreamOutput, onAfter);
             }
         }
         responseStatsConsumer.addResponseStats(message.length());
+        final var responseType = writeable.getClass();
+        final boolean compress = compressionScheme != null;
         internalSend(
             channel,
             message,
-            networkMessage,
+            requestAction == null
```

Comment:
As I said above, not 100% sure about this one, but then again, either we want this information or not though and this still seems cheaper than redundantly building the string.

Reply:
Yep I think we want this info. Looks easy enough to plumb in the `action` on the response path too these days, but we can do that in a followup.